src/main/kotlin/biokotlin/util/MergeGVCFUtil.kt

```diff
@@ -0,0 +1,361 @@
+        println("Time next lowest position: ${timeNextLowestPosition / 1e9} secs.")
+        println("Time lowest look ahead start: ${timeLowestLookAheadStart / 1e9} secs.")
+        println("Time result: ${timeResult / 1e9} secs.")
+        println("Time advance variant: ${timeAdvanceVariant / 1e9} secs.")
+
+    }
+
+    runBlocking {
+        writeOutputVCF(outputFile, samples, variantContextChannel)
+    }
+
+}
+
+private fun createVariantContext(
+    samples: List<String>,
+    variants: Array<SimpleVariant?>,
+    currentPosition: Position
+): VariantContext? {
+
+    val variantsAtPosition = variants
+        .filterNotNull()
+        .filter { it.positionRange.contains(currentPosition) }
+
+    val hasSNP = variantsAtPosition.find { it.isSNP } != null
+    val hasIndel = variantsAtPosition.find { it.isIndel } != null
+
+    // This is set up to handle SNPs that doesn't overlap with indels
+    // But can be expanded to handle other types of variants
+    return when {
+        hasSNP && !hasIndel -> snp(variants, currentPosition, samples)
+        else -> null
+    }
+
+}
+
+private suspend fun writeOutputVCF(
+    outputFile: String,
+    samples: List<String>,
+    variantContextChannel: Channel<Deferred<VariantContext?>>
+) {
+
+    // Write the merged VCF file, using the HTSJDK VariantContextWriterBuilder
+    VariantContextWriterBuilder()
+        .unsetOption(Options.INDEX_ON_THE_FLY)
+        .setOutputFile(File(outputFile))
+        .setOutputFileType(VariantContextWriterBuilder.OutputType.VCF)
+        .setOption(Options.ALLOW_MISSING_FIELDS_IN_HEADER)
+        .build()
+        .use { writer ->
+
+            val header = VCFHeader(createGenericVCFHeaders(samples))
+            writer.writeHeader(header)
+
+            var timeWriting = 0L
+            for (deferred in variantContextChannel) {
+                val variantContext = deferred.await()
+                measureNanoTime {
+                    if (variantContext != null) writer.add(variantContext)
+                }.let { timeWriting += it }
+            }
+
+            println("Time writing VCF: ${timeWriting / 1e9} secs.")
+
+        }
+
+}
+
+/**
+ * Creates a VariantContext for the current position, given the variants
+ * at that position from the GVCF readers.
+ */
+private fun snp(variants: Array<SimpleVariant?>, currentPosition: Position, samples: List<String>): VariantContext {
+
+    var refAllele: String? = null
+    var altAlleles: MutableSet<String> = mutableSetOf()
+    val variantsUsed = mutableListOf<SimpleVariant>()
+    val genotypes: List<Pair<Boolean, List<String>>> = variants
+        .map { variant ->
+            when (variant) {
+                null -> Pair(false, listOf(".")) // No call, since no variant at position for this sample
+                else -> {
+                    if (variant.positionRange.contains(currentPosition)) {
+
+                        val variantRef = when {
```

Comment:
I would recommend moving this chunk of code to something like the following:
`val variantRef = getVariantRef( ...)`

Then you can unit test this very easily by passing in different scenarios and make sure the logic is correct.

Reply:
Ok, did that